### PR TITLE
End to end diffing

### PR DESCRIPTION
**Pull request by michaelcfanning**
This change permits capturing user-facing output as part of test diffing.

This change isn't ideal/fully realized: it's authored in this haphazard way to permit an upstream submodule consumer to accept it.

After shipping in main, we should push these changes deeper into the core test API (ensuring that *this* repos tests continue to function).

### CODE REVIEW

File: src/Test.Utilities.Sarif/FileDiffingUnitTests.cs [hunk 1]
```diff
@@ -22,6 +22,8 @@
 
 using Xunit.Abstractions;
 
+using YamlDotNet.Core.Tokens;
```
Comment (shaopeng-gh) [inline src/Test.Utilities.Sarif/FileDiffingUnitTests.cs:25]: is this needed or can be deleted

File: src/Sarif/EnumeratedArtifact.cs [hunk 2]
```diff
@@ -19,131 +19,78 @@ public EnumeratedArtifact(IFileSystem fileSystem)
         internal byte[] bytes;
         internal string contents;
 
-        private Encoding encoding;
```
[the remark this answers is not in the document]
Comment (michaelcfanning) [reply src/Sarif/EnumeratedArtifact.cs:22]: Oops, should have cherry-picked this specific commit, fixing.

File: src/Sarif/EnumeratedArtifact.cs [hunk 3]
```diff
@@ -19,131 +19,78 @@ public EnumeratedArtifact(IFileSystem fileSystem)
...
-                    (Uri.IsAbsoluteUri && !Uri.IsFile))
-                {
-                    throw new InvalidOperationException("An absolute URI pointing to a file location was not available.");
-                }
-
-                // This is our client-side, disk-based file retrieval case.
-                this.Stream = FileSystem.FileOpenRead(Uri.LocalPath);
-            }
+                // TBD we actually have no validation URI is non-null yet.
+                contents = Uri!.IsFile
```
Comment (shaopeng-gh) [inline src/Sarif/EnumeratedArtifact.cs:51]: >IsFile [](http://example.com/codeflow?start=32&length=6)

not sure if applies here,
Uri.IsFile will throws an InvalidOperationException when it is a relative Uri

File: src/Sarif/EnumeratedArtifact.cs [hunk 4]
```diff
@@ -19,131 +19,78 @@ public EnumeratedArtifact(IFileSystem fileSystem)
...
 
-            if (Stream.CanSeek)
-            {
-                RetrieveDataFromSeekableStream();
+                this.sizeInBytes = (long?)this.contents?.Length;
             }
             else
             {
-                RetrieveDataFromNonSeekableStream();
+                if (Stream.CanSeek) { this.Stream.Seek(0, SeekOrigin.Begin); }
```
Comment (shaopeng-gh) [inline src/Sarif/EnumeratedArtifact.cs:59]: >Stream [](http://example.com/codeflow?start=20&length=6)

just a note the else part of the if logic itself, 
does not guarantee Stream is not null here

File: src/Sarif/EnumeratedArtifact.cs [hunk 5]
```diff
@@ -19,131 +19,78 @@ public EnumeratedArtifact(IFileSystem fileSystem)
...
+                this.sizeInBytes = Stream.Read(this.bytes, 0, this.bytes.Length);
+                Stream.Close();
+                Stream = null;
             }
+
+            return this.bytes;
         }
 
+
         public long? sizeInBytes;
```
Comment (shaopeng-gh) [inline src/Sarif/EnumeratedArtifact.cs:94]: extra empty line can be removed

**Pull request closed without merging**